api/src/main/java/org/apache/iceberg/RowKey.java

```diff
@@ -0,0 +1,213 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.List;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Types;
+
+/**
+ * Row key of a table.
+ * <p>
+ * Row key is a definition of table row uniqueness,
+ * similar to the concept of primary key in a relational database system.
+ * A row should be unique in a table based on the values of each {@link RowKeyField}.
+ * Iceberg itself does not enforce row uniqueness based on this key.
+ * It is leveraged by operations such as streaming upsert.
+ */
+public class RowKey implements Serializable {
```

Comment:
I don't have a strong opinion but I'd go for `RowIdentifier` or `RowId`. I think `Key` means uniqueness but I'll be fine this way too as long as we agree Iceberg does not ensure uniqueness.

Reply:
The `key` does not always means `uniqueness` in my mind.  from the MySQL [document](https://dev.mysql.com/doc/refman/8.0/en/create-index.html),  `index` could be created on top of key columns and the `index` could choose to be unique or non-unique.

Reply:
I think we have gone back and forth on this naming, and for now I would prefer the `Key` case because `Id` is heavily used in table metadata to mean concepts such as `spec-id`, `schema-id`, `order-id`, etc. which are the increasing ID of different specs. Using a different keyword `Key` would provide more clarity in the table metadata.

Reply:
I agree with Jack's logic that "id" is typically used in Iceberg to refer to a numeric identifier. It would be odd to use `RowId`, especially given the overlap with the JDBC one. But, we have had a significant number of people that find "key" confusing when it is a non-unique "key".

What about shifting the focus from the "key" or "identifier" to the fields? We could use `identity-field-ids` to hold the collection and add `identityFieldIds` to APIs.

Reply:
I renamed class `RowKeyField` to `RowKeyIdentifierField`, and `fields` to `identifier-fields` in metadata. Please let me know if that feels better.

Reply:
Personally, I don't have strong opinion about the `RowKeyField` or `RowKeyIdentifierField`.  I'm okay if you think it's good for one of them.

Reply:
To be more clear, I don't think we should ignore the near consensus from our sync discussion that "key" is misleading. I think we should instead call this class `IdentityFields` (or something similar) and store `identifier-field-ids` in table metadata.